README
README file for Atomic Tanks
============================================


What is Atomic Tanks?
========================

Atomic Tanks is a game in which you control an overly-powerful tank and attempt
to blow up other highly powerful tanks. Players get to select a number of
weapons and defensive items and then attack each other in a turn-based manner.
The last tank standing is the winner.



How to I get/install Atomic Tanks?
===================================

To gain a copy of Atomic Tanks, go to our website at
   http://atanks.sourceforge.net
Click on the "Download" link and select the latest version of Atanks.

Note: If you have a pre-compiled (binary) copy of Atomic Tanks, you will need
      the Allegro graphic package.
      See : http://allegro.cc/
      Also, if you are going to compile/build Atomic Tanks from source code, you
      will need the Allegro development package.

If you are using a pre-built, binary copy of Atomic Tanks, please see the next
section on running the game. Next we will cover compiling from source.


To compile Atomic Tanks, please make sure you have the Allegro libraries and
headers installed (see above).
Next, open a command console (shell) and move into the directory containing the
Atomic Tanks package. This file will probably be named "atanks-6.5.tar.gz".
Then run the following commands:

tar zxf atanks-6.5.tar.gz
cd atanks-6.5

Now it depends on your preferences:

1) Default linux installation :
The Binary goes to  : /usr/bin
The data files go to: /usr/share/games/atanks

make
make install

If you are on Ubuntu, or one of Ubuntu's children distributions, and encounter
sound problems, you can try to to fix this by building and installing Atanks
with:

make ubuntu
make install ubuntu

2) Linux User installation, where everything stays in the atanks folder :

make user

3) Windows User installation where everything stays in the atanks folder :

make winuser

4) OSX User installation where everything stays in the atanks folder :

make osxuser
(You may wish to use "gmake osxuser" instead.)


Now we can get to running the game. If you encounter an error while running
"make", then please see the last section of this document, which covers
reporting bugs.


How to adapt build and installation paths
=========================================

There are two different variables which influence how Atanks is built and
installed. Both are important to package managers in the first place:

PREFIX - This defaults to /usr and influences the basic path. If you would like
         to install into /usr/local instead, simply call:
         make PREFIX=/usr/local
         make install PREFIX=/usr/local

DESTDIR - This defaults to an empty string and can be set for build roots, just
          like the prefix. If you do not know what that means, you do not need
          it. ;-)


How do I run Atomic Tanks?
====================================

You can run the atanks file by typing "./atanks" from a command line, or by
double-clicking the file in a file browser.
Atomic Tanks tries to use reasonable defaults, so there is rarely a need to use
the command-line options. However, Atomic Tanks supports the following options:

-h  --help           Show this help screen
-fs                  Full screen
    --windowed       Run in a window
-w  --width <width>  Specify the screen width in pixels
-t  --tall  <height> Specify the screen height in pixels
                     Adjust the screen size at your own risk
                     (default is 800x600)
-d  --depth <depth>  Colour depths, currently either 16 or 32 bit
    --datadir <path> Path to the data directory
-c <path>            Path to config and saved game directory
    --noconfig       Do not load game settings.
    --nosound        Disable sound
    --noname         Do not show player name above tank
    --nonetwork      Disable network connections.
    --nobackground   Do not display the green menu background.

Most of these options can be used together, with the exception of "-fs" for full
screen mode. Simply running "./atanks" is the same as running Atomic Tanks with
the following options:
./atanks --windowed --width 800 --tall 600 --datadir . depth 32


How do I go about creating a player for myself?
================================================

when you first run Atomic Tanks, the first thing you'll need to do is to create
a player/tank for yourself. If you haven't created a human player, yet, the game
will present you with the player creation screen automatically. Otherwise click
on the "Players" button and click "Create New". Enter the name you want and pick
the colour you'd like your tank to be.
The teams "Jedi" and "Sith" mean that their members work (somewhat) together,
while the neutral tanks are all on their own. The "Jedi" AI is a bit more social
and defensive, while the "Sith" AI is a bit more individualistic and aggressive.

That's all you'll need to do. Click the "Okay" button. Then click the "Back"
button and, most importantly, click "Play"!

Once you click play, you'll have a chance to select which tanks you'll be
playing against. You can get anywhere from 2 to 10 tanks on the playing field at
one time. Don't forget to select yourself!

The next screen gives you a chance to buy weapons and defensive items. There is
no perfect combination, so experiment a little.
You can buy an item (assuming you have enough funds) by left-clicking on the
item. To sell the item, right-click on the item. When you've completed your
selection, click "Done".

The final screen you'll see is where the action takes place!
Use the left and right arrow buttons on your keyboard to aim your tank's gun.
The up and down arrow keys adjust the speed (power) of the shot. The space bar
fires the tanks gun. If you want to change which weapon you will be using, press
the TAB button or the BACKSPACE button. The round is over when there is one tank
left standing.


Keyboard Controls
=========================================

Atomic Tanks tries to be keyboard friendly and much of the tank's controls
require the keyboard. Here is a quick overview of which keys do what:

SPACEBAR   -- Fires weapons and selects/toggles menu items.
ENTER      -- Is similar to pressing the OK button on a menu.
UP/DOWN    -- Adjusts tank power, cycles through menu items and scrolls on the
              buying screen.
LEFT/RIGHT -- Aim the tank's gun. Also buys and sells on the buying screen or
              adjusts values in menus.
ESC        -- Cancel out of a menu.
F1         -- Takes a screenshot.
F10        -- Tells the computer to take over your tank for the remainder of the
              round.
              Also saves games when on the buying screen.
V, v       -- The "v" key controls the volume during matches. Pressing
              lower-case "v" decreases the volume and upper-case "V" increases
              the volume.
~          -- Show/hide the scoreboard during rounds. On german (and possibly
              other) keyboards the scoreboard is switched with the key '#'.


How to play a game over a network?
=========================================

Atomic Tanks has just introduced network play and it is still a bit rough.
However, if you'd like to try playing this way, here are a new points to help
you along.

Playing over a network requires two or more people. One person must be the Host
and the other person is the Client. If you are hosting the game, make sure
you have networking enabled. To check this, go into the Options menu and select
the Network option. Make sure "Networking" is set to "On". If it was "Off", then
change it to "On" and then exit Atanks and re-start the game. Congratulations,
you're now running a Host. You can continue to play as you normally would from
this point. Clients will connect to your game and take over computer-controled
tanks. To connect to you, the Clients will need your IP address. To get this,
visit the website whatismyip.com.

For clients, to connect to a host, you'll need to first get the IP address of
your Host. Once you have that, go to the Option menu and select the Network
option. Change the box marked "Server Address" to match the IP address of your
Host. Then return to the main menu and click the button labelled Network Game.
Atanks will then try to connect to the server and introduce you to the game.

Things are still flaky on the Client side. Hopefully it will improve as time
goes on.

If the Client cannot connect to the Host, there are a few possible problems.
Perhaps there is a firewall in the way? Maybe the Host is using a non-default
port number. Perhaps the Client side just isn't working well enough yet. If you
run into problems, please let us know about them on the forum.

When you're playing a game (as the client) you'll notice the various tanks are
in four different colours. The Jedi Team is marked in Green, the Sith are marked
as Purple, Neutral tanks are Blue and the player's tank is indicated with Red.
Hopefully this will prevent you from shooting your team mates.


Known issues
=========================================

There are some known issues which players should be aware of. We are trying to
fix these, but in the mean time, it's best to know they exist and how to work
around them. The issues are, in no particular order.

1. The default sound driver doesn't work on Ubuntu Linux and its related
distributions. If sound isn't working for you, go into the Options menu, select
Sound. Make sure Sound is turned on and that the Sound Driver is switched from
AutoDetect to OSS. Then return to the main menu, Quit and then re-start the
game. If this doesn't resolve the issue, then Quit the game and delete all of
the files and folders in your computer's /tmp folder with the name "pulse".

2. The client side of the network games is buggy. We're working on that.


Where do I report bugs or send requests?
=========================================

Atomic Tanks is a great game, but it isn't perfect.
We, the developers, are always trying to improve the experience our game brings
to you. If you have a problem with installing or playing Atomic Tanks, please
let us know.
We have a bug tracker on xxx, where issues can be recorded, tracked, and
discussed.
Or mail the project lead at jessefrgsmith@yahoo.ca
We encourage you to give your feedback. Please give as many details as possible
so we can help you.



Changelog
NOTE: From now on, new changes appear at the top of this file.

	Fixes:
	- Avoid divide by zero error during land calculation. Fix provided by Sven.
	- Assert is often called when attempting to draw virtual object without a
	  bitmap associated with the object. This has been removed to avoid exiting
	  the game and a simple error is printed to the terminal instead.


============ Atanks-6.6 released ===================

	Fixes:
	- Falling damage that was caused by a driller or riot weapon, is now
	  credited to the shooter of that weapon.
	- Fixed a crash that happened when SDI units tried to target meteors.
	- Fixed Makefile so that tar abd zip commands would not include
	  .git directory in source archives.
	Additions:
	- The AI players can move their tank to get to a better position.
	- New weapon: "Theft Bomb" (Idea and basic patch by Alois Dirnaichner)
	- Game save files got a version number, so old save games can be
	  upgraded.
	Changes:
	- Raised the power reduction of the Reducer from 25% to 33%.
	- The radius of the riot and dirt bombs have been raised.
	- The prices of some items and weapons have been rebalanced.
	- The height rollers can climb depends on their size now.
	- All players start with some fuel.
	- Human players are no longer considered to be like deadly bots, but
	  to be like the best present bot. In a game with only lower level
	  bots, they no longer "unite" against the human players right from
	  the start.
	- SDI units might skip their shot if the resulting explosion will kill
	  the SDI owning tank.
	- SDI units might skip their shot if the resulting explosion will be
	  nearer to the tank than on the point were the missile would explode
	  if not shot down.
	- Fixed bug in debug.h which would prevent compiling on FreeBSD.
	- Reduced AI difficultly, especially at lower levels. AI was far too
	  strong for most players. AI aiming now uses a number of aiming attempts
	  equal to its level rather than a power of its level.


============ Atanks-6.5 released ===================

	- Remove old Atanks binary during normal "make clean".

	- Added patch from Debian which makes the build process reproducible.
	  Thanks to Reiner Herrmann for submitting the patch. Closes
	  Debian bug #842865.


============ Atanks-6.4 released ===================

	Service release with bug fixes and some improvements for the AI.
	Fixes:
	- Fixed a bug that caused the AI to use Napalm in a stupid way.
	Changes:
	- The funky bombs are now a lot more 'funky' by randomly switching
	  horizontal and/or vertical direction in mid-flight.

============ Atanks-6.3 released ===================

	Sven has done a huge amount of code changes for this release. The
	key highlights are as follows:
	- Improved AI.
	- Smoother/faster buying screen navigation.
	- Huge re-work of the Options screen and other menus to
	  avoid crashes and random symbols appearing.
	- Improved demo mode.
	- The AI does a better job of buying items/weapons.
	- Updated Makefiles to reduce the number of separate Makefiles required.
	- Atanks should now build natively in Visual Studio on Windows.
	- Improved Unicode support.
	- Various code clean-up.
	- Added dirt debris.
	- Improved revenge mode for AI.
	- Menus should now respond faster.
	- Network code clean-up.
	- Added more in-code documentation.

============ Atanks-6.2 released ===================

	- Large code clean-up by Bruno Victal
	- More style/code clean-up by Bruno Victal
	    Indentation
	    Removing unnecessary braces
	    Removing unnecessary parentheses
	    Removing unnecessary spaces
	    Removing many blank lines (line wasting)
	    Expanding some ifs and switch cases

	- Bruno Victal provided further code clean-up
	    Renamed old referenced files in the perror calls (from .cc to .cpp)
	    Removed some useless casts in printf calls [ printf((char *)"String here"); ]
	    Deleted the macros about DOS support (they were already commented out)
	    Fixed abs() related math functions so they do not throw compiler warnings.

	- Removed extra call to Get_Team_Name() that was not required.

	Daniel sent in a patch which addresses the following:
	- fixes a bug where the address of (stack allocated) temporal arrays
	    was stored in the menu structure causing
	    wrong behavior, e.g. printing garbage in the Network configuration
	    menu title.
	- fixes const-correctness for all the menu structure
	- moves stack-allocated read-only arrays to .rodata so no large stack
	    memory is used.
	- adds Daniel to the list of people (feel free to remove me if this

	Bruno Victal provided two more fixes.
	- Fix isnan() ambiguity when compiled with C++11 standards
	- Removed OLD_GAMELOOP code and references. Stick with
	    modern game loop.


=========== Atanks-6.1 released ==============
	All of this releases patches were provided by Bill Buerger.
	- Extend the amount of time the AI will play against itself
	  from 10 seconds to 60 seconds.
	- Fixed "clean" command in Windows Makefile.
	- Players now wait to start their turn until they
	  can control their tank. No more waiting at the start
	  of their turn.
	- Rollers now properly explode when hitting steel wall.
	- Scoreboard shows current income and turn order.
	  This can be toggled with the ~ key.
	- Fixed navigation keys on the store/buying sceen.


=========== Atanks-6.0 released ==============
	- Removed extra alleg42.dll file from src directory.
        - Clarified license.
        - Pressing "v" or "V" during a match will cause
          the audio volume to decrease or increase, respectively.
          Volume patch provided by Daniel.
        - Plugged memory leak from update check.
        - Fixed possible memory problem in floating text.

=========== Atanks-5.9 released ==============

	- Applied patch which removes GCC warnings,
          Patch supplied by Daniel Gutson.


============ Atanks-5.8 released ==============

	- Removed compiler optimizations from src/Makefile to prevent

src/main.h
#ifndef MAIN_DEFINE
#define MAIN_DEFINE

/*
 * atanks - obliterate each other with oversize weapons
 * Copyright (C) 2003  Thomas Hudson
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 * */

#ifndef VERSION
#  error "VERSION information is missing. Fix Makefile."
#endif

#ifdef GENTOO
#  ifndef DATA_DIR
#    define DATA_DIR "/usr/share/games/atanks"
#  endif // DATA_DIR
#endif // GENTOO

#ifndef BUFFER_SIZE
# define BUFFER_SIZE 256
#endif


/// Important: debug.h not only detects which OS/compiler this is,
/// it puts an important fix with allegro on windows in place.
/// Therefore it *must* stay before the block including winalleg.h!
#include "debug.h"


// The windows port does some crazy stuff with int*_t types.
#if defined(ATANKS_IS_WINDOWS)
#  include <cstdint>
#  ifndef ALLEGRO_HAVE_STDINT_H
#    define ALLEGRO_HAVE_STDINT_H 1
#  endif // ALLEGRO_HAVE_STDINT_H
#  if !defined(ATANKS_SRC_ATANKS_CPP)
#    define ALLEGRO_NO_MAGIC_MAIN
#  endif // Not called from atanks.cpp
#endif // Windows build system

#  include <allegro.h>

#if defined(ATANKS_IS_WINDOWS)
#  include <winalleg.h>
#endif // windows


// For visual studio some "workarounds" must be put in place
#if defined(ATANKS_IS_MSVC)
#  define PATH_MAX MAX_PATH
   // Needed for M_PIl to show up
#  if !defined(_USE_MATH_DEFINES)
#    define _USE_MATH_DEFINES 1
#  endif
#  include <cmath>
#else
#  include <unistd.h>
#  include <cmath>
#endif // Windows versus Linux


// Be sure M_PI and M_PIl are set:
#if !defined(M_PI)
#  define M_PI 3.14159265358979323846f
#endif
#if !defined(M_PIl)
#  define M_PIl 3.14159265358979323846L
#endif


#include <cstdlib>
#include <cstdio>
#include <cstring>
#include <ctime>
#include <iostream>
#include <fstream>
#include <string>
#include <chrono>
#include <thread>
#include <algorithm>


#include "globaltypes.h"


#define	DONE_IMAGE             11
#define	FAST_UP_ARROW_IMAGE    12
#define	UP_ARROW_IMAGE         13
#define	DOWN_ARROW_IMAGE       14
#define	FAST_DOWN_ARROW_IMAGE  15


#ifndef HAS_DIRENT
#  if defined(ATANKS_IS_MSVC)
#    include "extern/dirent.h"
#  else
#    include <dirent.h>
#  endif // Linux
#  define HAS_DIRENT 1
#endif // HAS_DIRENT


// Some more workarounds to compile using visual studio:
#if defined(ATANKS_IS_MSVC)
#  define snprintf         atanks_snprintf
#  define strncpy(d, s, c) strncpy_s(d, c+1, s, c)
#  define strncat(d, s, c) strncat_s(d, c+1, s, c)
#  define sscanf           sscanf_s
#  define access           _access
#  define F_OK             02
#  define R_OK             04
#  define W_OK             02
#  define strcasecmp       _stricmp
#  define strdup           _strdup
#  define unlink           _unlink
#  define mkdir            _mkdir
#endif

// Note: See winclock.h why this is necessary
/// REMOVE_VS12_WORKAROUND
#if defined(ATANKS_IS_MSVC) && !defined(ATANKS_IS_AT_LEAST_MSVC13)
#define USLEEP(microseconds_) Sleep(microseconds_ / 1000);
#define MSLEEP(milliseconds_) Sleep(milliseconds_);
#else
#define USLEEP(microseconds_) std::this_thread::sleep_for(std::chrono::microseconds(microseconds_));
#define MSLEEP(milliseconds_) std::this_thread::sleep_for(std::chrono::milliseconds(milliseconds_));
#endif // VS12 workaround
#define LINUX_SLEEP     MSLEEP(10)
#define LINUX_REST      MSLEEP(40)


using std::cerr;
using std::cout;
using std::endl;
using std::string;


// place to save config and save games
#if defined(ATANKS_IS_WINDOWS)
#  define HOME_DIR "AppData"
#elif defined(ATANKS_IS_LINUX)
#  define HOME_DIR "HOME"
#endif // Windows versus Linux

#ifndef	DATA_DIR
#  define DATA_DIR "."
#endif

#define MAX_OVERSHOOT 10000


// The nex few are some math helpers that shorten things dramatically.
#define SIGN(x_arg)   ((x_arg) < 0  ? -1  : 1 )
#define SIGNd(x_arg)  ((x_arg) < 0. ? -1. : 1.)
#define ROUND(x_arg)  static_cast<int32_t>( (x_arg) + (SIGNd(x_arg) * .5))
#define ROUNDu(x_arg) static_cast<uint32_t>((x_arg) + .5)

#define FABSDISTANCE2(x1,y1,x2,y2) \
	std::sqrt(std::pow(static_cast<double>(x2) - static_cast<double>(x1), 2.) \
	        + std::pow(static_cast<double>(y2) - static_cast<double>(y1), 2.) )
#define FABSDISTANCE3(x1, y1, z1, x2, y2, z2) \
	std::sqrt(std::pow(static_cast<double>(x2) - static_cast<double>(x1), 2.) \
	        + std::pow(static_cast<double>(y2) - static_cast<double>(y1), 2.) \
	        + std::pow(static_cast<double>(z2) - static_cast<double>(z1), 2.) )

#define ABSDISTANCE2(x1,y1,x2,y2)       ROUNDu(FABSDISTANCE2(x1,y1,x2,y2))
#define ABSDISTANCE3(x1,y1,z1,x2,y2,z2) ROUNDu(FABSDISTANCE3(x1,y1,z1,x2,y2,z2))

#define DEG2RAD(degree_) (static_cast<double>(degree_) * M_PIl / 180. )
#define RAD2DEG(radian_) (static_cast<double>(radian_) * 180.  / M_PIl)


/** @brief show or hide the custom mouse cursor
  *
  * This macro can either hide the custom mouse cursor when @a where is set
  * to nullptr, or draw it on @a where, which then must be a pointer to a
  * BITMAP.
  *
  * This macro should be used to hide the custom mouse cursor before doing any
  * drawing and to place the mouse cursor on @a where once all other drawing is
  * done.
  *
  * If the OS mouse cursor is used, this macro does nothing.
  *
  * @param[in] where BITMAP pointer to draw the custom cursor on or nullptr to
  * hide the custom mouse cursor.
**/
#define SHOW_MOUSE(where) { \
	if (!env.osMouse) { \
		if (where != nullptr) unscare_mouse(); \
		else scare_mouse(); \
		show_mouse(where); \
		/* Make the neccessary updates */ \
		if (where != nullptr) { \
			global.make_update (mouse_x, mouse_y, env.misc[0]->w, env.misc[0]->h); \
			global.make_update (lx, ly, env.misc[0]->w, env.misc[0]->h); \
			lx = mouse_x; \
			ly = mouse_y; \
		} \
	} \
}

#define MAXPLAYERS    10
#define MAX_POWER   2000
#define MIN_POWER    100
#define MAX_ROUNDS 10000

#define MENUHEIGHT 40
#define BOXED_TOP 41 // This is the highest non-border pixel in boxed mode
#define	BALLISTICS 53
#define	BEAMWEAPONS 3
#define WEAPONS (BALLISTICS + BEAMWEAPONS)
#define ITEMS 24
#define THINGS (WEAPONS + ITEMS)
#define	NATURALS 6
#define DIRT_FRAGMENT -1
#define RADII 6
#define MAXRADIUS 200
#define BUTTONFRAMES 2

#define MENUBUTTONS 7
#define INGAMEBUTTONS 4
#define SPREAD 10
#define NAME_LEN 24
#define ADDRESS_LENGTH 16

#define WAIT_AT_END_OF_ROUND 1 // second (enough with the new live score board)

#define MAX_ITEM_NAME_LEN 127
#define MAX_ITEM_DESC_LEN 511
#define MAX_ITEMS_IN_STOCK 999999
#define MAX_MONEY_IN_WALLET 1000000000

// to make the theft bomb base steal easier to change, here
// is a useful define. Maybe, one day, we'll add it to the options?
#define THEFT_AMOUNT 5000

// Use these instead of the (most strict) defaults,
// But only where timing by memory fences do not matter.
#define ATOMIC_READ  std::memory_order_acquire
#define ATOMIC_WRITE std::memory_order_release


//turns
enum turnTypes
{
	TURN_HIGH = 0,
	TURN_LOW,
	TURN_RANDOM,
	TURN_SIMUL
};


struct POINT_t
{
	int32_t x = 0;
	int32_t y = 0;

	explicit
	POINT_t() { }
	POINT_t(int32_t x_, int32_t y_);
	POINT_t &operator=( const POINT_t &src );
};


struct BOX
{
	int32_t x = 0;
	int32_t y = 0;
	int32_t w = 0;
	int32_t h = 0;

	explicit
	BOX () { }
	BOX (int32_t x_, int32_t y_, int32_t w_, int32_t h_);
	BOX &operator=(const BOX  &src);
	BOX &operator=(const BOX &&src);

	void set(int32_t x_, int32_t y_, int32_t w_, int32_t h_);
};

// Make the BOX usage easier:
bool operator==(const BOX &lhs, const BOX &rhs);
bool operator!=(const BOX &lhs, const BOX &rhs);


struct gradient
{
	RGB color;
	float point;
};


class WEAPON
{
public:

	/* -----------------------------------
	 * --- Constructors and destructor ---
	 * -----------------------------------
	 */

	explicit WEAPON();


	/* -----------------------------------
	 * --- Public methods              ---
	 * -----------------------------------
	 */

	int32_t     getDelayDiv() const;
	const char* getDesc()     const;
	const char* getName()     const;

	void setDesc(const char* desc_);
	void setName(const char* name_);


	/* -----------------------------------
	 * --- Public members              ---
	 * -----------------------------------
	 */
	int32_t cost            = 0;  //!< $ :))
	int32_t amt             = 0;  //!< number of weapons in one buying package
	double  mass            = 0.;
	double  drag            = 0.;
	int32_t radius          = 0;  //!< of the explosion
	int32_t sound           = 0;
	int32_t etime           = 0;
	int32_t damage          = 0;  //!< damage power
	int32_t picpoint        = 0;  //!< which picture do we show in flight?
	int32_t spread          = 0;  //!< number of weapons in the shot
	int32_t delay           = 0;  //!< volleys etc.
	int32_t noimpact        = 0;
	int32_t techLevel       = 0;
	int32_t warhead         = 0;  //!< Is it a warhead?
	int32_t numSubmunitions = 0;  //!< Number of submunitions
	int32_t submunition     = 0;  //!< The next stage
	double  impartVelocity  = 0.; //!< Impart velocity 0.0-1.0 to subs
	int32_t divergence      = 0;  //!< Total angle for submunition spread
	double  spreadVariation = 0.; //!< Uniform or random distribution
	                              //!< 0-1.0 (0=uniform, 1.0=random)
	                              //!< divergence at centre of range
	double  launchSpeed     = 0.; //!< Speed given to submunitions
	double  speedVariation  = 0.; //!< Uniform or random speed
	                              //!< 0-1.0 (0=uniform, 1.0=random)
	                              //!< launchSpeed at centre of range
	int32_t countdown       = 0;  //!< Set the countdown to this
	double  countVariation  = 0.; //!< Uniform or random countdown
	                              //!< 0-1.0 (0=uniform, 1.0=random)
	                              //!< countdown at centre of range


private:


	/* -----------------------------------
	 * --- Private members             ---
	 * -----------------------------------
	 */

	char desc[MAX_ITEM_DESC_LEN + 1];
	char name[MAX_ITEM_NAME_LEN + 1];
};

#define	MAX_ITEMVALS	10
class ITEM
{
public:

	/* -----------------------------------
	 * --- Constructors and destructor ---
	 * -----------------------------------
	 */

	explicit ITEM();


	/* -----------------------------------
	 * --- Public methods              ---
	 * -----------------------------------
	 */

	const char* getDesc() const;
	const char* getName() const;

	void setDesc(const char* desc_);
	void setName(const char* name_);


	/* -----------------------------------
	 * --- Public members              ---
	 * -----------------------------------
	 */

	int32_t cost       = 0;
	int32_t amt        = 0;
	int32_t selectable = 0;
	int32_t techLevel  = 0;
	int32_t sound      = 0;
	double  vals[MAX_ITEMVALS];


private:


	/* -----------------------------------
	 * --- Private members             ---
	 * -----------------------------------
	 */

	char desc[MAX_ITEM_DESC_LEN + 1];
	char name[MAX_ITEM_NAME_LEN + 1];
};


enum shieldVals
{
	SHIELD_ENERGY,
	SHIELD_REPULSION,
	SHIELD_RED,
	SHIELD_GREEN,
	SHIELD_BLUE,
	SHIELD_THICKNESS
};


enum selfDestructVals
{
	SELFD_TYPE = 0,
	SELFD_NUMBER
};


enum weaponType
{
	SML_MIS            =  0,
	MED_MIS            =  1,
	LRG_MIS            =  2,
	SML_NUKE           =  3,
	NUKE               =  4,
	DTH_HEAD           =  5,
	SML_SPREAD         =  6,
	MED_SPREAD         =  7,
	LRG_SPREAD         =  8,
	SUP_SPREAD         =  9,
	DTH_SPREAD         = 10,
	ARMAGEDDON         = 11,
	CHAIN_MISSILE      = 12,
	CHAIN_GUN          = 13,
	JACK_HAMMER        = 14,
	SHAPED_CHARGE      = 15,
	WIDE_BOY           = 16,
	CUTTER             = 17,
	SML_ROLLER         = 18,
	LRG_ROLLER         = 19,
	DTH_ROLLER         = 20,
	SMALL_MIRV         = 21,
	ARMOUR_PIERCING    = 22,
	CLUSTER            = 23,
	SUP_CLUSTER        = 24,
	FUNKY_BOMB         = 25,
	FUNKY_DEATH        = 26,
	FUNKY_BOMBLET      = 27,
	FUNKY_DEATHLET     = 28,
	BOMBLET            = 29,
	SUP_BOMBLET        = 30,
	BURROWER           = 31,
	PENETRATOR         = 32,
	SML_NAPALM         = 33,
	MED_NAPALM         = 34,
	LRG_NAPALM         = 35,
	NAPALM_JELLY       = 36,
	DRILLER            = 37,
	TREMOR             = 38,
	SHOCKWAVE          = 39,
	TECTONIC           = 40,
	RIOT_BOMB          = 41,
	HVY_RIOT_BOMB      = 42,
	RIOT_CHARGE        = 43,
	RIOT_BLAST         = 44,
	DIRT_BALL          = 45,
	LRG_DIRT_BALL      = 46,
	SUP_DIRT_BALL      = 47,
	SMALL_DIRT_SPREAD  = 48,
	CLUSTER_MIRV       = 49,
	PERCENT_BOMB       = 50,
	REDUCER            = 51,
	THEFT_BOMB         = 52, // Last ballistic (BALLISTICS == 53)
	SML_LAZER          = 53,
	MED_LAZER          = 54,
	LRG_LAZER          = 55, // Last weapon (WEAPONS == 56)
	SML_METEOR         = 56,
	MED_METEOR         = 57,
	LRG_METEOR         = 58,
	SML_LIGHTNING      = 59,
	MED_LIGHTNING      = 60,
	LRG_LIGHTNING      = 61 // Last natural
};


#define LAST_EXPLOSIVE DRILLER

#define ITEM_NO_SHIELD   -1
enum itemType
{
	ITEM_TELEPORT             =  0, // 56 (weap_idx - WEAPONS)
	ITEM_SWAPPER              =  1, // 57
	ITEM_MASS_TELEPORT        =  2, // 58
	ITEM_FAN                  =  3, // 59
	ITEM_VENGEANCE            =  4, // 60
	ITEM_DYING_WRATH          =  5, // 61
	ITEM_FATAL_FURY           =  6, // 62
	ITEM_LGT_SHIELD           =  7,
	ITEM_MED_SHIELD           =  8,
	ITEM_HVY_SHIELD           =  9,
	ITEM_LGT_REPULSOR_SHIELD  = 10,
	ITEM_MED_REPULSOR_SHIELD  = 11,
	ITEM_HVY_REPULSOR_SHIELD  = 12,
	ITEM_ARMOUR               = 13,
	ITEM_PLASTEEL             = 14,
	ITEM_INTENSITY_AMP        = 15,
	ITEM_VIOLENT_FORCE        = 16,
	ITEM_SLICKP               = 17,
	ITEM_DIMPLEP              = 18,
	ITEM_PARACHUTE            = 19,
	ITEM_REPAIRKIT            = 20,
	ITEM_FUEL                 = 21, // 77
	ITEM_ROCKET               = 22, // 78
	ITEM_SDI                  = 23  // 79 (Last item)
};

#define SHIELD_COUNT		6

//signals
#define	SIG_QUIT_GAME         -1
#define SIG_OK                 0
#define GLOBAL_COMMAND_QUIT   -1
#define GLOBAL_COMMAND_MENU    0
#define GLOBAL_COMMAND_OPTIONS 1
#define GLOBAL_COMMAND_PLAYERS 2
#define GLOBAL_COMMAND_CREDITS 3
#define	GLOBAL_COMMAND_HELP    4
#define GLOBAL_COMMAND_PLAY    5
#define GLOBAL_COMMAND_DEMO    6
#define GLOBAL_COMMAND_NETWORK 7


/** @enum eClasses
  * @brief class definitions of everything from virtual objects up
  *
  * The ordering here determines the order of the drawing.
**/
enum eClasses
{
	CLASS_MISSILE = 0,
	CLASS_BEAM,
	CLASS_TANK,
	CLASS_TELEPORT,
	CLASS_DECOR_DIRT,
	CLASS_DECOR_SMOKE,
	CLASS_EXPLOSION,
	CLASS_FLOATTEXT,
	CLASS_COUNT
};


#ifndef HAS_TANK
class TANK; // forwarding if not known
#endif // HAS_TANK

/// === Global functions used in several compilation units ====
void   drawMenuBackground(eBackgroundTypes backType, int32_t tOffset,
                          int32_t numItems);
double interpolate       (double x1, double x2, double i);
double Noise             (int x);
double Noise2D           (int x, int y);
double perlin1DPoint     (double amplitude, double scale, double xo,
                          double lambda, int octaves);
double perlin2DPoint     (double amplitude, double scale, double xo, double yo,
                          double lambda, int octaves);
void   quickChange       (bool clearerror);

#include "externs.h"

#endif

